rust/scx_rustland_core/assets/bpf/main.bpf.c

```diff
@@ -783,6 +857,49 @@ void BPF_STRUCT_OPS(rustland_exit_task, struct task_struct *p,
 	__sync_fetch_and_add(&nr_queued, 1);
 }
 
+/*
+ * Check if we can switch to FIFO mode if the system is underutilized.
+ *
+ * We consider the system underutilized if the amount of tasks that are running
+ * or waiting to run is less than half the amount of CPUs available in the
+ * system.
+ */
+static bool should_enable_fifo(void)
+{
+	/* Moving average of the currently active tasks */
+	static u64 nr_active_avg;
+	u64 nr_active;
+
+	if (!fifo_sched)
+		return false;
+
+	/* Get the amount of active tasks (running or waiting to run) */
+	nr_active = nr_running + nr_queued + nr_scheduled;
+
+	/* Update moving average of active tasks */
+	nr_active_avg = (nr_active_avg + nr_active) / 2;
+
+	/*
+	 * We should react fast at exiting from FIFO mode, to prevent potential
+	 * stuttering behavior is the system becomes suddenly overloaded.
+	 *
+	 * At the same time, we should be more careful at entering FIFO mode,
+	 * because, in this case, we may introduce stuttering behavior if the
+	 * system is overloaded and we don't properly detect it (false
+	 * positive), entering, incorrectly, in FIFO mode.
+	 *
+	 * In order to mitigate the false positives, use a "smoother" moving
+	 * average for the amount of active tasks to determine if we need to
+	 * enter FIFO mode, and use the last sample of the active tasks to
+	 * determine if we need to exit FIFO mode.
+	 */
+	if (!is_fifo_enabled)
+		nr_active = nr_active_avg;
+
+	/* Update flag that determines if FIFO needs to be enabled */
+	return nr_active <= (num_possible_cpus >> 1);
```

Comment:
I wonder whether having separate enable and disable thresholds would be useful so that a system which is sitting right on the edge doesn't keep flipping between the two modes. Or maybe that's okay?

Reply:
Hm... good point. There's the moving average that is trying to smooth `nr_active` a bit, maybe adding also a separate threshold can help to reduce the risk of flipping between the two modes.

Also, flipping between the two modes is not dramatic, the downside is that "FIFO" tasks will get more priority than the "non-FIFO" tasks, because the former are dispatched using `SCX_DSQ_LOCAL`. Now that I think about it, that's not really needed, actually they should be dispatched to the shared DSQ. I'll run some tests with this change.

Reply:
Actually no... I take it back, using the shared DSQ would be bad, because they could be bounced on the first available CPU, instead of running them on their local one, so `SCX_DSQ_LOCAL` is better to keep them running on the CPU selected during `select_cpu`.

Reply:
> Edit: ...but, we can use the per-CPU DSQ (without using `dispatch_task()`, because we can't apply the whole `scx_bpf_dispatch_cancel()` logic in `enqueue`).

Basically 4b2defce6ca3edb8bc247035720274b4c02441ed, that I like it better. than using SCX_DSQ_LOCAL and it also seems to perform slightly better.